staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/evaluator.go

```diff
@@ -0,0 +1,237 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/generic"
+)
+
+// evaluatorCompiler implement the interface EvaluatorCompiler.
+type evaluatorCompiler struct {
+}
+
+func NewEvaluatorCompiler() EvaluatorCompiler {
+	return &evaluatorCompiler{}
+}
+
+type validationActivation struct {
+	object, oldObject, params, request interface{}
+}
+
+// ResolveName returns a value from the activation by qualified name, or false if the name
+// could not be found.
+func (a *validationActivation) ResolveName(name string) (interface{}, bool) {
+	switch name {
+	case ObjectVarName:
+		return a.object, true
+	case OldObjectVarName:
+		return a.oldObject, true
+	case ParamsVarName:
+		return a.params, true
+	case RequestVarName:
+		return a.request, true
+	default:
+		return nil, false
+	}
+}
+
+// Parent returns the parent of the current activation, may be nil.
+// If non-nil, the parent will be searched during resolve calls.
+func (a *validationActivation) Parent() interpreter.Activation {
+	return nil
+}
+
+// Compile compiles the cel expressions defined in the CELValidations into an Evaluator
+func (c *evaluatorCompiler) Compile(validations []Validation, hasParam bool) Evaluator {
+	if len(validations) == 0 {
+		return nil
+	}
+	compilationResults := make([]CompilationResult, len(validations))
+	for i, validation := range validations {
+		compilationResults[i] = CompileValidatingPolicyExpression(validation.Expression, hasParam)
+	}
+	return NewEvaluator(compilationResults, validations)
+}
+
+// evaluator implements the Evaluator interface
+type evaluator struct {
+	compilationResults []CompilationResult
+	validations        []Validation
+}
+
+func NewEvaluator(compilationResults []CompilationResult, validations []Validation) Evaluator {
+	return &evaluator{
+		compilationResults,
+		validations,
+	}
+}
+
+func convertObjectToUnstructured(obj interface{}) (*unstructured.Unstructured, error) {
+	if obj == nil || reflect.ValueOf(obj).IsNil() {
+		return &unstructured.Unstructured{Object: nil}, nil
+	}
+	ret, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	return &unstructured.Unstructured{Object: ret}, nil
+}
+
+func objectToResolveVal(r runtime.Object) (interface{}, error) {
+	if r == nil || reflect.ValueOf(r).IsNil() {
+		return nil, nil
+	}
+	v, err := convertObjectToUnstructured(r)
+	if err != nil {
+		return nil, err
+	}
+	return v.Object, nil
+}
+
+// Evaluate evaluates the compiled CEL expressions converting them into CELEvaluations
+// errors per evaluation are returned on the Evaluation object
+func (v *evaluator) Evaluate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) ([]Evaluation, error) {
+	// TODO: replace unstructured with ref.Val for CEL variables when native type support is available
```

Comment:
I think native type support has merged (https://github.com/kubernetes/kubernetes/pull/113312), but I also don't think it's a big deal if we merge without fixing this TODO here.

Reply:
cc @jiahuif

Reply:
@DangerOnTheRanger will unstructured still be supported once that's integrated? Evaluating policy shift-left means that we don't necessarily know the schema (CRD or built-in b/c version skew)

Reply:
It looks like the UnstructuredToVal() function handles the unknown schema use-case?

Reply:
It will not affect by type checking. Nothing changed from evaluation side hence we will still be using dyn type.